### PR TITLE
feat: add gem fiat ramp integration

### LINKED ISSUE #847: Gem Integration and User Flow
# Background
As part of building out our newly launched application (app.shapeshift.com) we want to integrate a fiat on-ramp to allow users to directly purchase crypto.  This is an important feature to allow us to achieve parity with our old application (beta.shapeshift.com) where we support Gem. 

We are looking for a skilled bounty hunter who wants to take on this larger feature and see it through to release.

Our engineering work-stream is actively hiring for full time roles, and this would be a great introduction if the bounty hunter is interested.  If not, that is of course okay as well, we welcome all contributions!

# User Stories

- As a user I want to buy/sell crypto to/from fiat on app.shapeshift.com
- As a DAO we want to earn revenue every time someone buys/sells with Gem.

# Design Mocks
 - [See Sketch Artboard for all mocks a few samples are below](https://www.sketch.com/s/839fdf8f-8022-4042-9767-a8a28951c847/a/yg4qWvo)
  
Screen Shot 2022-01-25 at 7 05 39 AM
Screen Shot 2022-01-25 at 7 05 49 AM
Screen Shot 2022-01-25 at 7 05 59 AM
Screen Shot 2022-01-25 at 7 06 13 AM

# User Experience & Flow

1. **Buy/Sell Button** - As a user, I want to be taken to a buy/sell screen from anywhere on app.shapeshift.com.
    1. Buy sell button should be next to wallet name in top bar.
2. **Select GEM Button** - As a user, I want to be able to select Gem as my fiat ramp provider from a list of providers
    1. Also include a button for onJuno with message “coming soon”
3. **Select "Buy" or "Sell" From Tabs -** As a user, I want to toggle between buy and sell depending on which action I want to do.
    1. If "**Buy**" is selected a dropdown should indicate all the possible assets I can buy.
        1. User should now be able to select the desired asset from a list.
        2. User selects desired account from a list of accounts connected in current wallet. If there is only one account in wallet, that account is selected by default.
        3. Desired asset address is passed through as a parameter from ShapeShift Web to Gem
        4. If user is using a hardware wallet, they have the option to select “confirm on wallet” button to confirm address shown on ShapeShift matches the address shown on their hardware wallet.
        5. Gem Interface launches in new browser tab or window.
        6. No further action required from ShapeShift Web.
    2. If "**Sell**" is selected a dropdown should indicate all the possible assets I can sell.
        1. All assets available to sell should be listed sorted by balance...assets with 0 balance listed last
        2. User should select an asset and see a list of accounts that hold said asset in their wallet. 
        3. If a user holds an asset in more than 1 account, interface gives them the option to select any desired account.
        4. User should select the account with that asset that they wish to sell from
        5. If user is using a hardware wallet, they have the option to select “confirm on wallet” button to confirm address shown on ShapeShift matches the address shown on their hardware wallet.
        6. Gem Interface launches and the user will enter the amount to send into Gem.
        7. User will copy the receiving address and amount from Gem and input into the Send module on ShapeShift Web.
        8. No further action required from ShapeShift Web.

# Revenue Share Technical Details

- In order for Gem to recognize traffic coming from ShapeShift, implementation w**i**ll Pass “Partner Name” to GEM to get Rev share back.

# Acceptance Criteria

- Entire gem workflow is built out as per mocks and the above user flow.
- All user inputs are validated correctly and ENS addresses are functional for ETH based transactions
- Bounty hunter will coordinate with our operations team who will assist in testing in a feature branch and fix any issues with their assistance. 
- Bounty hunter will follow up on any code review comments or suggestions from the engineering work-stream

# Bounty Hunters 
- Please only apply to work if you have fully read this entire specification, have time to dedicate to working on this immediately, and will be able to communicate with our internal work streams effectively.  We hope to complete this entire project within the next 3 weeks.
- Join our discord [here](https://discord.gg/shapeshift). Once you join you can see more about this issue [here](https://discord.com/channels/554694662431178782/873244924026449940/925799058537336853)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info

# Documentation and helpful references
- [Example Code](https://shapeshift.notion.site/Gem-Example-Code-1-266151e772104e0290c8633e2ecd1461)
- [https://api.gem.co/institutions/coinify/supported_currencies](https://api.gem.co/institutions/coinify/supported_currencies)
- [https://api.gem.co/institutions/wyre/supported_currencies](https://api.gem.co/institutions/wyre/supported_currencies)

*Note: Use [beta.shapeshift.com](http://beta.shapeshift.com) for user flows

Product Manager: @decentralizedj in Discord. Please reach out with any questions.

### PR DESCRIPTION
**Pull request by visitskyworld**
## Description
- Added `BuySellModal`
- Added `GemManager` for gem integration.
- User can toggle between buy and sell depending on which action he wants to do
- If "Buy" is selected, a dropdown should indicate all the possible assets that can be bought.
- If "Sell" is selected, a dropdown should indicate all the possible assets that can be sold sorted by balance.
- Redirect to Gem Interface if user clicks `Continue` button.

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #847  

## Risks
`selectPortfolioCryptoHumanBalanceBySymbol`  might cause collisions with assets that have the same symbol, but are actually different ones on the same chain, or even on a totally different chain.

## Screenshots (if applicable)
Screenshot_75
Screenshot_76
Screenshot_77
Screenshot_78
Screenshot_79
Screenshot_80

### CODE REVIEW

File: src/AppProviders.tsx [hunk 1]
```diff
@@ -58,7 +59,9 @@ export function AppProviders({ children }: ProvidersProps) {
                       <MarketDataProvider>
                         <TransactionsProvider>
                           <ModalProvider>
-                            <DefiProvider>{children}</DefiProvider>
+                            <BuySellManagerProvider>
+                              <DefiProvider>{children}</DefiProvider>
```
Comment (gomesalexandre) [inline src/AppProviders.tsx:63]: Disclaimer: I didn't start the proper review of this yet, so I might come back to what I said later on, take this with a grain of salt 🧂!

I feared this would potentially be used, since I made the same "mistake" (again, pinch of salt until I have a more in-depth look) while starting work on modals. tl;dr, `<DefiProvider />` is exclusively for the `DeFi` section and shouldn't be used to create modals.
It has very specific routing/logic and overall is a non-conventional modal for us (using a mix of controlled/uncontrolled logic etc).
`useModal()` should be used to consume modals, and their definition should live in`src/context/ModalProvider/ModalProvider.tsx`:
https://github.com/shapeshift/web/blob/e22521102e767406bca821d3dad025ec31bcfc39/src/context/ModalProvider/ModalProvider.tsx#L14

tl;dr 
- We don't need another provider for this, `ModalProvider` should do the trick and using a `<MemoryRouter />` should solve your modal routing needs
- `<DefiProvider />`shouldn't be used for anything else than DeFi context

Comment (visitskyworld) [reply src/AppProviders.tsx:63]: Thanks for your comment!

**Comment by 0xdef1cafe** [comment]
thanks for the PR @visitskyworld!

let me get all the environment variables added to CI tomorrow so we can see the tests passing.

@purelycrickets heads up this will be coming ops way tomorrow for an initial pass of testing

File: src/components/Modals/BuySell/providers/GemManager.tsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+const GEM_URL = getConfig().REACT_APP_GEM_URL
+
+export const GemManager = () => {
+  const translate = useTranslate()
+  const toast = useToast()
+  const { buysell } = useModal()
+
+  const [asset, setAsset] = useState<BuySellAsset | null>()
+  const [isSelectingAsset, setIsSelectingAsset] = useState(false)
+  const [verified, setVerified] = useState<boolean | null>(null)
```
Comment (gomesalexandre) [inline src/components/Modals/BuySell/providers/GemManager.tsx:43]: Do we actually use `verified` anywhere apart from button state/styling? From the ACC, this should allow to verify the address on a hardware wallet

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Thanks for the contribution @visitskyworld!

First review, I'll make a second review more focused on the business logic tomorrow.
This one was mostly about incorrect/missing typings, naming improvements/conventions, and testing. A lot of comments, but most of them are just repetitions!

In regards the actual testing of this, I've noticed that:
~~1. The "Sell" button on the initial modal routes to the buy modal~~
2. The styles in general are off compared to the mockups, with missing/incorrect positioning/distances, paddings and sizes

I am not able to test the hardware wallet part but I added a comment there also

File: src/components/Modals/BuySell/components/AssetSearch/AssetSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    arg.transaction_direction === 'bank_blockchain' ||
+    arg.transaction_direction === 'card_blockchain'
+
+  // Filter for the assets you can sell
+  const sellFilter = (arg: { transaction_direction: string }) =>
+    arg.transaction_direction === 'blockchain_bank'
+
+  //Filter and merge function
+  const filterAndMerge = useMemo(
+    () => (arr1: any[], arr2: any[], key: string | number, filter: any) => {
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/Modals/BuySell/components/AssetSearch/AssetSearch.tsx]: This is still typed as `any[]`

File: src/components/Modals/BuySell/components/AssetSearch/AssetSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  // Filter for the assets you can sell
+  const sellFilter = (asset: { transaction_direction: string }) =>
+    asset.transaction_direction === 'blockchain_bank'
+
+  //Filter and merge function
+  const filterAndMerge = useMemo(
+    () => (coinifyList: any[], wyreList: any[], key: string | number, filter: any) => {
+      const list1 = coinifyList
+        .filter(filter)
+        .map((list: { [x: string]: { currencies: any } }) => list[key].currencies)
```
Comment (gomesalexandre) [inline src/components/Modals/BuySell/components/AssetSearch/AssetSearch.tsx:66]: If filter functions are typed properly, `list` becomes typed implicitly so there will be no need for typing.

And again, if having abstracted types (`type x = {}`) and reusing them, you shouldn't have to type this with an incomplete subset of the full type of the response from `REACT_APP_GEM_COINIFY_SUPPORTED_COINS`
[Typescript Playground - Demo](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKhDOwoF4oG8BQVtQGYHt8AuKAcgCMBDAJ1IwF8MMBjfAO0SmAWBLkQDaAXRRQBmHHkIkKNOvSFNWHJLgCWAG27UAYm1EAKAB58eAShQA+KACFCGiJTbGzS9p26IYAd3yjPwAB06loQ1AYh2npmgQC2lGDGVkZmQA)

File: src/components/Modals/BuySell/providers/GemManager.tsx [hunk 5]
```diff
@@ -0,0 +1,194 @@
+import { ArrowBackIcon, CheckIcon, ChevronRightIcon, CopyIcon, ViewIcon } from '@chakra-ui/icons'
```
Comment (gomesalexandre) [inline src/components/Modals/BuySell/providers/GemManager.tsx:1]: Same comment about this living in `/providers`, this is not a context provider.
What about `ramp(s)` naming for everything currently named `provider(s)`?

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Regarding the typings, one major improvement that will help you get everything properly typed would be to add a return type for `getCoinifySupportedCurrencies` and `getWyreSupportedCurrencies`.
You can keep the existing `BuySellAsset` for the items in `destination.currencies` and `source.currencies`.
Once that's done, fixing all the type comments should be a breeze!

File: src/components/Modals/BuySell/components/AssetSearch/AssetSearch.tsx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  const getWyreSupportedCurrencies = async () => {
+    try {
+      const { data } = await axios.get(getConfig().REACT_APP_GEM_WYRE_SUPPORTED_COINS)
+      return data
+    } catch (ex: any) {
+      console.error(ex)
+    }
+  }
+  //Filter for the assets you can buy
+  const buyFilter = (arg: { transaction_direction: string }) =>
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/Modals/BuySell/components/AssetSearch/AssetSearch.tsx]: Kind of fixed, but that brought new concerns, look at the newer comment (https://github.com/shapeshift/web/pull/1062#discussion_r814613926)

File: src/components/Modals/BuySell/components/AssetSearch/AssetSearch.tsx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  //Filter for the assets you can buy
+  const buyFilter = (asset: { transaction_direction: string }) =>
+    asset.transaction_direction === 'bank_blockchain' ||
+    asset.transaction_direction === 'card_blockchain'
+
+  // Filter for the assets you can sell
+  const sellFilter = (asset: { transaction_direction: string }) =>
+    asset.transaction_direction === 'blockchain_bank'
+
+  //Filter and merge function
```
Comment (gomesalexandre) [inline src/components/Modals/BuySell/components/AssetSearch/AssetSearch.tsx:61]: It's called `filterAndMerge` so it's understandable that it is a `Filter and merge function` by the naming itself 😁

**Comment by visitskyworld** [comment]
@gomesalexandre , Thanks for your review.
Regarding the asset typings, I copied the sample code provided here: https://shapeshift.notion.site/Gem-Example-Code-1-266151e772104e0290c8633e2ecd1461
It would be great if you update the doc. Or do you want me to update the doc as well?

**Comment by gomesalexandre** [comment]
> @gomesalexandre , Thanks for your review. Regarding the asset typings, I copied the sample code provided here: https://shapeshift.notion.site/Gem-Example-Code-1-266151e772104e0290c8633e2ecd1461 It would be great if you update the doc. Or do you want me to update the doc as well?

I don't think examples are expected to be complete and include the whole implementation with TS, they should serve as examples. - they are probably copied from the https://beta.shapeshift.com implementation and serve as a reference on how to build the Gem request.

**Comment by gomesalexandre** [comment]
A few other things I've noticed that aren't here according to the ACC:
- `All assets available to sell should be listed sorted by balance...assets with 0 balance listed last` - assets are sorted in the received order from Gem, independently of wallet holdings
- `User should select an asset and see a list of accounts that hold said asset in their wallet.` - this isn't implemented
- `If a user holds an asset in more than 1 account, interface gives them the option to select any desired account.` - ditto
- `All user inputs are validated correctly and ENS addresses are functional for ETH based transactions` - I see no explicit support for this, but I wrote to @decentralized-j to double-check if ENS should be displayed instead of address for consistency

**Comment by visitskyworld** [comment]
Thanks, @gomesalexandre .
>If a user holds an asset in more than 1 account, interface gives them the option to select any desired account.
I am not sure if user can connect with multiple account, can I get [beta.shape](https://beta.shapeshift.com/) code so I can reference?

File: src/assets/translations/messages-en.json [hunk 8]
```diff
@@ -57,6 +60,25 @@
...
+    "gem": "GEM",
+    "gemMessage": "Buy or Sell Crypto with Gem",
+    "onJuno": "OnJuno",
+    "comingSoon": "Coming Soon ...",
+    "buy": "Buy",
+    "sell": "Sell",
+    "assetToBuy": "Assest to Buy",
+    "selectAnAssestToBuy": "Select an asset to buy",
+    "assetToSell": "Assest to Sell",
+    "selectAnAssestToSell": "Select an asset to sell",
```
Comment (NeOMakinG) [inline src/assets/translations/messages-en.json:77]:
```suggestion
    "assetToBuy": "Asset to Buy",
    "selectAnAssetToBuy": "Select an asset to buy",
    "assetToSell": "Asset to Sell",
    "selectAnAssetToSell": "Select an asset to sell",
```

It's a typo, right?

Comment (NeOMakinG) [reply src/assets/translations/messages-en.json:77]: Don't forget to change every uses of these keys!

**Comment by gomesalexandre** [comment]
> Thanks, @gomesalexandre .
> 
> > If a user holds an asset in more than 1 account, interface gives them the option to select any desired account.
> 
> I am not sure if user can connect with multiple account, can I get [beta.shape](https://beta.shapeshift.com/) code so I can reference?

Yeah, that was a mixup in ACC wording, apologies for that. What we meant here is actually "asset", the app doesn't support connecting multiple accounts, even with native, so that would be something impossible to do.
Regarding ENS, I got confirmation from @decentralized-j it should be displayed in place of the address if there's a lookup available for an address.

**Comment by 0xdef1cafe** [comment]
hey @visitskyworld can you please rebase off of develop - i added environment variables to CI so your tests will have a chance to pass

**Comment by 0xdef1cafe** [comment]
> 

@visitskyworld - do you have an ETA for this one? we'd like to keep this moving along.

**Comment by gomesalexandre** [comment]
> @0xdef1cafe @gomesalexandre , I will finish it asap. However I have some questions regarding your comment 
> Is this for hardware wallet address verification? If so, how can I get `chainAdapter`? I found code here but I am not sure what `asset` should be.
> 
> ```
>   const { chain, name, symbol } = asset
>   const chainAdapter = chainAdapterManager.byChain(chain)
> ```

This isn't related to hardware wallets or any type of wallet in particular. What we would like here is feature parity with the existing ENS flows.
If you haven't tried the flow yet, you can open the app and start sending funds to some ENS address or alternatively, receive (if you have an ENS for your current account). Notice how the address gets replaced with the resolved ENS name whenever applicable.

From a code perspective, you can also search for `ensReverseLookup` in the codebase. Again, the expected flow is that wherever you have an address visible, if there is an ENS resolution for it, it should be shown in place of the address (without breaking anything in the actual request-building logic).


> @0xdef1cafe @gomesalexandre , I will finish it asap. However I have some questions regarding your comment @gomesalexandre
> 
> > User should select an asset and see a list of accounts that hold said asset in their wallet.
> 
> I think user can connect with only one account and I am not sure how the user see a list of accounts.

Please have a look at my previous comment, there was some product wording that caused confusion.
What we mean here is not "an account", but "an asset" e.g BNB or USDC. @0xean @decentralized-j can we get this updated on the issue as well for clarity?

**Comment by decentralized-j** [comment]
> > @0xdef1cafe @gomesalexandre , I will finish it asap. However I have some questions regarding your comment @gomesalexandre
> > > User should select an asset and see a list of accounts that hold said asset in their wallet.
> > 
> > 
> > I think user can connect with only one account and I am not sure how the user see a list of accounts.
> 
> Please have a look at my previous comment, there was some product wording that caused confusion. What we mean here is not "an account", but "an asset" e.g BNB or USDC. @0xean @decentralized-j can we get this updated on the issue as well for clarity?

Yeah - chiming in here to clear it up, i think it was just a misunderstanding with phrasing. We want to display all of the assets in a given account, but there's no way to view all assets in all accounts at one time. If the user wants to see assets in a different account, they'll have to switch on their own.

phrased differently - we want to display all assets supported by shapeshift in the active account. no need to try and get multiple accounts all hooked up at once.

**Comment by gomesalexandre** [comment]
There is no `BuySellManagerProvider` - please fix the diff or edit the issue description.

File: src/components/Modals/BuySell/components/AssetSearch/AssetList.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import { AssetRow } from './AssetRow'
+
+type AssetListProps = {
+  handleClick: (asset: CurrencyAsset) => void
+  assets: CurrencyAsset[]
+  type: BuySellAction
+} & ListProps
+
+export const AssetList = ({ assets, type, handleClick }: AssetListProps) => {
+  return assets?.length === 0 ? (
```
Comment (gomesalexandre) [inline src/components/Modals/BuySell/components/AssetSearch/AssetList.tsx:15]: According to your typings, you don't need the `?` since the array can be empty, not undefined and `length` property exists on the array prototype i.e all arrays, even empty ones.
What about using `assets.length` and inverting the ternary?

File: src/components/Modals/BuySell/ramps/BuySellRamps.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+              height='auto'
+              justifyContent='flex-start'
+              variant='ghost'
+              fontWeight='normal'
+              py={2}
+            >
+              <Flex flexDirection='row' justifyContent='center' alignItems='center'>
+                <AssetIcon src={onjunologo} />
+                <Box textAlign='left' ml={2}>
+                  <Text fontWeight='bold' color='white' translation='buysell.onJuno' />
```
Comment (gomesalexandre) [inline src/components/Modals/BuySell/ramps/BuySellRamps.tsx:61]:
```suggestion
                  <Text fontWeight='bold' translation='buysell.onJuno' />
```

You can remove `color='white'` - it currently makes the text unreadable in light mode. Not using any value here will make it the right color depending on the color mode.

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
That looks way better from a code and typings perspective now!

One more thing I've also noticed, is that the way the balances are sorted isn't consistent with e.g Dashboard. They are sorted by crypto holdings amount, but for consistency with the rest of the app, we should probably use fiat amount (cc @decentralized-j).

image

As you can see here, my USDC is first, but my ETH is actually worth more in FIAT than all the others combined!

**Comment by gomesalexandre** [comment]
Also, could you have a look at the failing Cypress tests @visitskyworld ? Merge/rebase usually helps, but you might have to fix some tests to make it happy.

File: src/components/Modals/BuySell/ramps/GemManager.tsx [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+                      variant='ghost'
+                      onClick={copyHandler}
+                    />
+                    {!(wallet?.getVendor() === 'Native') ? (
+                      <IconButton
+                        icon={verified ? <CheckIcon /> : <ViewIcon />}
+                        onClick={handleVerify}
+                        aria-label='check-icon'
+                        size='sm'
+                        color={verified ? 'green.500' : verified === false ? 'red.500' : 'gray.500'}
```
[the remark this answers is not in the document]
Comment (visitskyworld) [reply src/components/Modals/BuySell/ramps/GemManager.tsx]: We have three states of `Verified`
- verify success
- verify failed
- not verified yet
> don't use this nested ternary altogether if possible.

How can I change this?

File: src/components/Modals/BuySell/ramps/GemManager.tsx [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+import { useModal } from 'context/ModalProvider/ModalProvider'
+import { useWallet } from 'context/WalletProvider/WalletProvider'
+import { ensReverseLookup } from 'lib/ens'
+
+import { BuySellAction, CurrencyAsset } from '../BuySell'
+import { AssetSearch } from '../components/AssetSearch/AssetSearch'
+import { getAssetLogoUrl } from '../components/AssetSearch/helpers/getAssetLogoUrl'
+import { BuySellActionButtons } from '../components/BuySellActionButtons'
+
+const middleEllipsis = (address: string, cut: number) =>
```
[the remark this answers is not in the document]
Comment (visitskyworld) [reply src/components/Modals/BuySell/ramps/GemManager.tsx:32]: Good point. Just wondering, can we use `<MiddleEllipsis />` inside `<Input />` element?
What about fixing like this?
`ensAddress || middleEllipsis(address as string, 11)`

**Comment by 0xdef1cafe** [comment]
> Details

cypress isn't running for outside contributors due to the way github manages secrets, we might have to reopen this under our org when it's ready for a final review

File: src/state/slices/portfolioSlice/selectors.ts [hunk 13]
```diff
@@ -196,6 +196,24 @@ export const selectPortfolioCryptoHumanBalanceByAssetId = createSelector(
     fromBaseUnit(bnOrZero(balances[assetId]), assets[assetId]?.precision ?? 0)
 )
 
+export const selectPortfolioCryptoHumanBalanceBySymbol = createSelector(
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/selectors.ts:199]: That works since gem doesn't use [CAIPs](https://github.com/shapeshift/caips) to identify assets, but please add it to the risks sections of your PR.
Identifying assets like this might cause collisions with assets that have the same symbol, but are actually different ones on the same chain, or even on a totally different chain.

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
From logic perspective, it mostly looks good now! A few smaller issues before giving this the big green stamp: 

1. If it would be possible and a low value for you to add some logic so that assets don't get re-fetched every time a user clicks the select asset to buy/sell button, it would be great. I'm happy if you just create a follow-up issue for it though
2. Styles still aren't matching the mocks:
  - On the "Buy or Sell Crypto" step, there should be a space between the Buy and Sell tags, and the left text which isn't present. As you add this, you will also notice that the modal is too narrow and should be wider
  
image


  - On the next step, padding is quite off. There's not enough on the x axis and too much padding-top. It becomes fine when you are in the "Select an asset to buy" step
  
image

  
  - Buy and Sell routing buttons sizes are off. They have more padding width and possibly more padding height on the design
  
image

  - Continue and cancel buttons are off. They should be according to design, which is also the same size as buttons in existing modals in the app e.g send modal
  
image

Feel free to approximate units, but make sure that it is as close as possible visually to design.

**Comment by visitskyworld** [comment]
@gomesalexandre I have fixed style and made assets not get re-fetched every time. PTAL.

Screenshot_87
Screenshot_88

**Comment by gomesalexandre** [comment]
What about Bitcoin @visitskyworld ? That uses ETH address in "Fund will be sent to", but actually, there is no Bitcoin support in Metamask I tested with. It should display the Bitcoin address - and the option to actually select bitcoin should be only for wallets that support it (KK/Native) (cc @decentralized-j should it be grayed out or filtered altogether from the list)?

image

**Comment by visitskyworld** [comment]
@gomesalexandre  We can not identify it now because gem doesn't support chain type.
Possible solution could be implementing `selectChainTypeBySymbol` similar to `selectPortfolioCryptoHumanBalanceBySymbol ` but it might be risky as you said.

**Comment by gomesalexandre** [comment]
> @gomesalexandre We can not identify it now because gem doesn't support chain type. Possible solution could be implementing `selectChainTypeBySymbol` similar to `selectPortfolioCryptoHumanBalanceBySymbol ` but it might be risky as you said.

What about simply using Bitcoin ChainType for Bitcoin, and Ethereum for other assets? On Gem, Bitcoin refers to native BTC in the Bitcoin network.

**Comment by visitskyworld** [comment]
Well, if gem supports cosmos or other chain types, what will we do?

**Comment by gomesalexandre** [comment]
> Well, if gem supports cosmos or other chain types, what will we do?

That's probably out of the scope of this issue, but that's great thinking forward!

**Comment by gomesalexandre** [comment]
Thanks for the work on this @visitskyworld !
As we've discussed in DM, we are closing this in favour of the reopened #1128 in the org, where we will take over.

Don't forget to get in touch with @0xean with the spent hours so we can compensate you appropriately!.
